### PR TITLE
FIX: Refactor assign / reassign to combine logic

### PR DESCRIPTION
**Pull request by janzenisaac**
Fixes: https://dev.discourse.org/t/self-assign-is-not-working-on-mobile/56331/21?u=isaacjanzen

- Combines assign and reassign logic... removing bulky and extra logic

### CODE REVIEW

**Comment by janzenisaac** [comment]
This feels like a flaky test... @jjaffeux What do you think?

**Comment by jjaffeux** [comment]
> This feels like a flaky test... @jjaffeux What do you think?

I think this is an issue given it's happening all the time :po

File: app/controllers/discourse_assign/assign_controller.rb [hunk 1]
```diff
@@ -44,11 +44,27 @@ def unassign
...
+      assign_to = username.present? ? User.find_by(username_lower: username.downcase) : Group.where("LOWER(name) = ?", group_name.downcase).first
+
+      raise Discourse::NotFound unless assign_to
+      raise Discourse::NotFound if !Assignment.valid_type?(target_type)
+      target = target_type.constantize.where(id: target_id).first
+      raise Discourse::NotFound unless target
+
+      # perhaps?
+      #Scheduler::Defer.later "assign topic" do
+      assign = Assigner.new(target, current_user).send("assign", assign_to)
```
Comment (ZogStriP) [inline app/controllers/discourse_assign/assign_controller.rb:61]: Why are you using `send` when you can call the `assign` method directly?

```suggestion
      assign = Assigner.new(target, current_user).assign(assign_to)
```

File: app/controllers/discourse_assign/assign_controller.rb [hunk 2]
```diff
@@ -44,11 +44,27 @@ def unassign
...
-      reassign_or_assign_target(action: "reassign")
+      assign_to = username.present? ? User.find_by(username_lower: username.downcase) : Group.where("LOWER(name) = ?", group_name.downcase).first
+
+      raise Discourse::NotFound unless assign_to
+      raise Discourse::NotFound if !Assignment.valid_type?(target_type)
+      target = target_type.constantize.where(id: target_id).first
+      raise Discourse::NotFound unless target
+
+      # perhaps?
+      #Scheduler::Defer.later "assign topic" do
```
Comment (ZogStriP) [inline app/controllers/discourse_assign/assign_controller.rb:60]: Is the `assign` method a CPU/DB intensive process?

File: app/controllers/discourse_assign/assign_controller.rb (hunk 1 above)
Comment (janzenisaac) [reply app/controllers/discourse_assign/assign_controller.rb:61]: This was left over from an old method. Good catch, I will remove.

File: app/controllers/discourse_assign/assign_controller.rb (hunk 2 above)
Comment (janzenisaac) [reply app/controllers/discourse_assign/assign_controller.rb:60]: From looking at the logs on assign, it does not seem overly heavy... this was a comment (left over) from before I started working on the re-assign workflow, so i figured I would leave it in. In the case someone was coming back to update? 🤷‍♂️

Comment (ZogStriP) [reply app/controllers/discourse_assign/assign_controller.rb:60]: I'd rather we remove the comment to keep the code clean (it'll be in the git history anyways). We can always revisit this later if we end up hitting a performance wall.

**Comment by janzenisaac** [comment]
@ZogStriP could i get a re-review on this?

**Review by eviltrout — APPROVED** [review]

**Pull request merged**